src/main/scala/com/daml/projection/JdbcAction.scala

```diff
@@ -162,42 +166,69 @@ object Sql {
-      case x: Short                     => ps.setShort(pos, x)
-      case x: Int                       => ps.setInt(pos, x)
-      case x: Long                      => ps.setLong(pos, x)
-      case x: Float                     => ps.setFloat(pos, x)
-      case x: Double                    => ps.setDouble(pos, x)
-      case x: BigDecimal                => ps.setBigDecimal(pos, x.bigDecimal)
-      case x: java.math.BigDecimal      => ps.setBigDecimal(pos, x)
-      case x: String                    => ps.setString(pos, x)
-      case x: java.sql.Date             => ps.setDate(pos, x)
-      case x: java.sql.Timestamp        => ps.setTimestamp(pos, x)
-      case x: java.sql.Array            => ps.setArray(pos, x)
-      case null                         => ps.setNull(pos, java.sql.Types.NULL)
-      case x: Optional[_] if x.isEmpty  => ps.setNull(pos, java.sql.Types.NULL)
-      case x: Optional[_] if !x.isEmpty => ps.setObject(pos, x.get())
-      // TODO fix in typeclass solution
-      case x: Option[_] if x.isEmpty => ps.setNull(pos, java.sql.Types.NULL)
-      case Some(x)                   => ps.setObject(pos, x)
-      case x                         => ps.setObject(pos, x)
-    }
-  }
+@FunctionalInterface
+trait Bind[-T] {
+
+  /** Sets the value on the [[java.sql.PreparedStatement]] */
+  def set(ps: PreparedStatement, pos: Int, value: T): Unit
+
+  /** Creates a [[Bind.Applied]] function that will set `value` at `pos` on a `PreparedStatement` argument. */
+  final def apply(value: T, pos: Int): Bind.Applied = set(_, pos, value)
 }
 
-/** Sets a bind value on a PreparedStatement */
-@SerialVersionUID(1L)
-@FunctionalInterface
-trait Setter extends java.io.Serializable {
+/**
+ * Provides `Bind` values for [[ExecuteUpdate.bind[T](pos*]],[[ExecuteUpdate.bind[T](name*]], [[Binder.bind[T](pos*]]
+ * and [[Binder.bind[T](name*]] methods.
+ */
+object Bind {
+  type Used[-T] = Bind[_ >: T]
+
+  private def mk[T <: AnyVal](f: (PreparedStatement, Int, T) => Unit): Bind[T] = new Bind[T] {
+    override def set(ps: PreparedStatement, pos: Int, value: T) = f(ps, pos, value)
+  }
+
+  private def mkN[T >: Null <: AnyRef](f: (PreparedStatement, Int, T) => Unit): Bind[T] = new Bind[T] {
+    override def set(ps: PreparedStatement, pos: Int, value: T) = {
+      if (value eq null) ps.setNull(pos, java.sql.Types.NULL) else f(ps, pos, value)
+    }
+    // TODO SC would a specific thing from Types be better?
+  }
 
-  /** Sets a bind value on `ps` */
-  def set(ps: java.sql.PreparedStatement): Unit
+  implicit val Boolean: Bind[Boolean] = mk(_.setBoolean(_, _))
+  implicit val Byte: Bind[Byte] = mk(_.setByte(_, _))
+  implicit val Short: Bind[Short] = mk(_.setShort(_, _))
+  implicit val Int: Bind[Int] = mk(_.setInt(_, _))
+  implicit val Long: Bind[Long] = mk(_.setLong(_, _))
+  implicit val Float: Bind[Float] = mk(_.setFloat(_, _))
+  implicit val Double: Bind[Double] = mk(_.setDouble(_, _))
```

Comment:
Java turns all of these into `Bind<Object>`, which unfortunately lets you do

```java
.bind(2, 42L, Bind.Int())
```

I think we need Java API counterparts with explicitly-written boxed types, i.e. `Bind[java.lang.Byte]` &c. And those should probably be the ones with the pleasant names, reserving funny space names for Scala.

Reply:
In Scala the implicits should obviate the need to use the 'unpleasant' name so this makes sense I think.

Reply:
Done.